### ISSUE TITLE
[inductor][accuracy] Not likely a bug but worth discussing - higher compute dtype precision

### ISSUE THREAD

**Issue by anijain2305** [pytorch/torchdynamo#1934]
Repro 

~~~
import torch._inductor.overrides

import torch
from torch import tensor, device
import torch.fx as fx
from torch._dynamo.testing import rand_strided
from math import inf
from torch.fx.experimental.proxy_tensor import make_fx

# REPLACEABLE COMMENT FOR TESTING PURPOSES

# torch version: 1.14.0a0+git2c0bd85
# torch cuda version: 11.6
# torch git version: 2c0bd85c755043d696452ddab354f3ff6775738b


# CUDA Info:
# nvcc: NVIDIA (R) Cuda compiler driver
# Copyright (c) 2005-2022 NVIDIA Corporation
# Built on Thu_Feb_10_18:23:41_PST_2022
# Cuda compilation tools, release 11.6, V11.6.112
# Build cuda_11.6.r11.6/compiler.30978841_0

# GPU Hardware Info:
# NVIDIA A100-SXM4-40GB : 8


from torch.nn import *
class Repro(torch.nn.Module):
    def __init__(self):
        super().__init__()



    def forward(self, wt_fp32, t_fp16, t_fp32):
        cast_fp16 = torch.ops.aten._to_copy.default(t_fp32, dtype = torch.float16)
        x_fp16 = torch.ops.aten.add.Tensor(cast_fp16, t_fp16)
        wp_fp16 = torch.ops.aten._to_copy.default(wt_fp32, dtype = torch.float16)
        conv = torch.ops.aten.convolution.default(x_fp16, wp_fp16, None, [1, 1], [0, 0], [1, 1], False, [0, 0], 1)
        return (conv,)

args = [((1152, 192, 1, 1), (192, 1, 1, 1), torch.float32, 'cuda'), ((2, 192, 7, 7), (9408, 49, 7, 1), torch.float16, 'cuda'), ((2, 192, 7, 7), (9408, 49, 7, 1), torch.float32, 'cuda')]
args = [rand_strided(sh, st, dt, dev) for (sh, st, dt, dev) in args]
mod = make_fx(Repro())(*args)

from torch._inductor.compile_fx import compile_fx_inner
from torch._dynamo.debug_utils import same_two_models

compiled = compile_fx_inner(mod, args)
class AccuracyError(Exception):
    pass
if not same_two_models(mod, compiled, args, only_fwd=True):
    raise AccuracyError("Bad accuracy detected")
~~~

This one fails accuracy. Since we fallback to `aten.convolution`, the the error arises from the `_to_copy` and `add` ops. 
Inductor uses higher compute dtype for pointwise ops, therefore even `_to_copy` first casts up to `float32` and downcasts to `float16` using tl.store.

Just doing this in Inductor produces accurate result 

~~~
diff --git a/torch/_inductor/codegen/triton.py b/torch/_inductor/codegen/triton.py
index de84d4ddbef..5873908ae25 100644
--- a/torch/_inductor/codegen/triton.py
+++ b/torch/_inductor/codegen/triton.py
@@ -94,9 +94,9 @@ def triton_compute_type(dtype):
     triton_type_name = str(dtype).split(".")[-1]
     if triton_type_name == "bool":
         triton_type_name = "int1"
-    if triton_type_name in ("float16", "bfloat16"):
-        # float16 math is done in float32 inside the kernel
-        triton_type_name = "float32"
+    # if triton_type_name in ("float16", "bfloat16"):
+    #     # float16 math is done in float32 inside the kernel
+    #     triton_type_name = "float32"
     return f"tl.{triton_type_name}"

~~~

Note that in this example, we can't compare with fp64 goldens because of an unrelated factor. It is possible that Inductor is closer to fp64 results.

But even then, it might be worth discussing if we expect the above repro to give exact same results as eager.

**Comment by anijain2305** [comment]
cc @ngimel 

Background, I am trying to debug and looking at the accuracy minified repros one by one
![image](https://user-images.githubusercontent.com/13822661/204223765-08ab19bc-b8e0-450c-9717-d754fa1adaff.png)

**Comment by ngimel** [comment]
No, we don't expect the above repro to give exact same results, for the reasons you pointed out. We do fewer intermediate truncations in inductor.

**Comment by anijain2305** [comment]
Cool. Closing this then.
